src/aolayer.cpp

```diff
@@ -544,19 +545,21 @@ void AOLayer::movie_ticker()
     else
       frame = 0;
   }
-  //  qint64 difference = elapsed - movie_delays[frame];
-  if (frame >= movie_frames.size()) {
-    movie_frames.append(this->get_pixmap(m_reader.read()));
-    movie_delays.append(m_reader.nextImageDelay());
-  }
-
+  future.waitForFinished(); // don't set the frame before we definitely have it in memory
 #ifdef DEBUG_MOVIE
   qDebug() << frame << movie_delays[frame]
            << "actual time taken from last frame:" << actual_time.restart();
 #endif
-
   this->set_frame(movie_frames[frame]);
   ticker->setInterval(this->get_frame_delay(movie_delays[frame]));
+  if (frame + 1 >= movie_frames.size() && frame + 1 < max_frames) { // load the next frame before we tick again
+      future = QtConcurrent::run(this, &AOLayer::load_next_frame);
```

Comment:
no waiting to finish?

this could cause a race condition if the frame hasn't finished reading yet and movie_ticker moves to the next frame and then tries to load another frame seeing as the next frame isn't ready yet.

also remember that a lot of these data structures aren't thread safe so you really want to make sure that two load_next_frame calls aren't happening at the same time

Reply:
> if i personally implemented this, what i would do is let the separate thread run all the way and read all of the frames into memory (or up to a certain # of frames ahead of the cursor).

the original implementation of this did load all frames into memory, which causes large amounts of lag at the start of the animation

I can for sure make this thread-safe though

Reply:
> since to avoid crashes the ticker still has to wait for the thread to finish

yeah you can also make this implementation thread safe by using a monitor (which is the traditional CS way to solve the producer-consumer problem). a monitor is basically a mutex + wait condition. there are some tutorials here:

- general tutorial on monitors https://doc.qt.io/archives/qq/qq21-monitors.html
- bounded buffer example https://doc.qt.io/qt-5/qtcore-threads-waitconditions-example.html
- you can also do it safely with semaphores instead of monitors in this case https://doc.qt.io/qt-5/qtcore-threads-semaphores-example.html
  however, monitors are still considered the higher-level and therefore superior construct.
  ever done auto-complete in a java IDE and seen those useless `notify()` and `wait()` methods? well it's because every `Object` in java is also a monitor. hope you learned something today

Reply:
I don't really want to preload more than one frame because I'm worried about it causing slowdown on high-framerate animations (think 60fps, you really don't have a lot of time to load stuff in between frames)

with that in mind, re-requesting review

Reply:
you misunderstand me - you are running it on a separate thread, and the main thread can take the first ready frame even before the decoding thread is done decoding all frames